### ISSUE TITLE
predictor.explain() not working with inputs of type "$()"

### ISSUE THREAD

**Issue by dipyamanroy** [amaiya/ktrain#518]
### Summary
I was using DistilBERT to train a model that parses a given request based on whether it is some malicious request or non-malicious, multi-class classification. Despite `predictor.predict()` working fine for the erroneous requests discussed below, `predictor.explain()` fails for query of type `$()`.

#### Error
![image](https://github.com/amaiya/ktrain/assets/92980894/17733113-27b6-43a1-9bc0-b916d4e87ac5)

#### Correction
However, as soon as either the brackets or dollar sign is removed, it works perfectly:
![image](https://github.com/amaiya/ktrain/assets/92980894/ff45c66e-c62c-4c0d-9c99-2df9673d78a0)
![image](https://github.com/amaiya/ktrain/assets/92980894/570f5f0c-dd40-4389-a573-c2834e5a3946)

Please assist on how to resolve this, as my data may contain queries of type `$()`.

**Comment by amaiya** [comment]
As shown below, I'm not able to reproduce your error on Google Colab (or locally):

![image](https://github.com/amaiya/ktrain/assets/47191980/6ef6a193-33d9-4b20-8b60-b923a27bac59)

If you share a Google Colab notebook reproducing the error, I can look it into it.

**Comment by dipyamanroy** [comment]
I'll try to share a notebook recreating it, but in the meantime, here's the full error output if it helps you decipher the issue:

```py
File ~\.conda\envs\BERT\lib\site-packages\ktrain\text\predictor.py:151, in TextPredictor.explain(self, doc, truncate_len, all_targets, n_samples)
    149 doc = " ".join(doc.split()[:truncate_len])
    150 te = TextExplainer(random_state=42, n_samples=n_samples)
--> 151 _ = te.fit(doc, self.predict_proba)
    152 return te.show_prediction(
    153     target_names=self.preproc.get_classes(), targets=prediction
    154 )

File ~\.conda\envs\BERT\lib\site-packages\eli5\lime\lime.py:261, in TextExplainer.fit(self, doc, predict_proba)
    258 self.y_proba_ = predict_proba(samples)
    259 self.clf_ = clone(self.clf)
--> 261 self.metrics_ = _train_local_classifier(
    262     estimator=self.clf_,
    263     samples=X,
    264     similarity=sims,
    265     y_proba=self.y_proba_,
    266     expand_factor=self.expand_factor,
    267     random_state=self.rng_
    268 )
    269 return self

File ~\.conda\envs\BERT\lib\site-packages\eli5\lime\lime.py:366, in _train_local_classifier(estimator, samples, similarity, y_proba, expand_factor, test_size, random_state)
    347 (X_train, X_test,
    348  similarity_train, similarity_test,
    349  y_proba_train, y_proba_test) = train_test_split(samples,
   (...)
    352                                                  test_size=test_size,
    353                                                  random_state=rng)
    355 # XXX: in the original lime code instead of a probabilitsic classifier
    356 # they build several regression models which try to output probabilities.
    357 #
   (...)
    364 # Ideally, it should be supported as a Pipeline (i.e. user should
    365 # be able to configure it).
--> 366 fit_proba(estimator, X_train, y_proba_train,
    367           expand_factor=expand_factor,
    368           sample_weight=similarity_train,
    369           random_state=rng)
    371 y_proba_test_pred = estimator.predict_proba(X_test)
    372 if y_proba_test_pred.shape != y_proba_test.shape:
    373     # Sometimes generated training labels may contain only a subset of
    374     # target classes; it means it could happen that dimensions
   (...)
    377     # XXX: the fix is not complete; to explain predictions
    378     # of the fitted estimator one still have to take care of target_names.

File ~\.conda\envs\BERT\lib\site-packages\eli5\lime\utils.py:37, in fit_proba(clf, X, y_proba, expand_factor, sample_weight, shuffle, random_state, **fit_params)
     30 X, y, sample_weight = expanded_X_y_sample_weights(X, y_proba,
     31     expand_factor=expand_factor,
     32     sample_weight=sample_weight,
     33     shuffle=shuffle,
     34     random_state=random_state,
     35 )
     36 fit_params = with_sample_weight(clf, sample_weight, fit_params)
---> 37 clf.fit(X, y, **fit_params)
     38 return clf

File ~\.conda\envs\BERT\lib\site-packages\sklearn\base.py:1152, in _fit_context.<locals>.decorator.<locals>.wrapper(estimator, *args, **kwargs)
   1145     estimator._validate_params()
   1147 with config_context(
   1148     skip_parameter_validation=(
   1149         prefer_skip_nested_validation or global_skip_validation
   1150     )
   1151 ):
-> 1152     return fit_method(estimator, *args, **kwargs)

File ~\.conda\envs\BERT\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:907, in BaseSGDClassifier.fit(self, X, y, coef_init, intercept_init, sample_weight)
    878 """Fit linear model with Stochastic Gradient Descent.
    879 
    880 Parameters
   (...)
    903     Returns an instance of self.
    904 """
    905 self._more_validate_params()
--> 907 return self._fit(
    908     X,
    909     y,
    910     alpha=self.alpha,
    911     C=1.0,
    912     loss=self.loss,
    913     learning_rate=self.learning_rate,
    914     coef_init=coef_init,
    915     intercept_init=intercept_init,
    916     sample_weight=sample_weight,
    917 )

File ~\.conda\envs\BERT\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:694, in BaseSGDClassifier._fit(self, X, y, alpha, C, loss, learning_rate, coef_init, intercept_init, sample_weight)
    691 # Clear iteration count for multiple call to fit.
    692 self.t_ = 1.0
--> 694 self._partial_fit(
    695     X,
    696     y,
    697     alpha,
    698     C,
    699     loss,
    700     learning_rate,
    701     self.max_iter,
    702     classes,
    703     sample_weight,
    704     coef_init,
    705     intercept_init,
    706 )
    708 if (
    709     self.tol is not None
    710     and self.tol > -np.inf
    711     and self.n_iter_ == self.max_iter
    712 ):
    713     warnings.warn(
    714         (
    715             "Maximum number of iteration reached before "
   (...)
    719         ConvergenceWarning,
    720     )

File ~\.conda\envs\BERT\lib\site-packages\sklearn\linear_model\_stochastic_gradient.py:648, in BaseSGDClassifier._partial_fit(self, X, y, alpha, C, loss, learning_rate, max_iter, classes, sample_weight, coef_init, intercept_init)
    638     self._fit_binary(
    639         X,
    640         y,
   (...)
    645         max_iter=max_iter,
    646     )
    647 else:
--> 648     raise ValueError(
    649         "The number of classes has to be greater than one; got %d class"
    650         % n_classes
    651     )
    653 return self

ValueError: The number of classes has to be greater than one; got 1 class
```

I apologize that my responses are delayed, we're probably in different timezones.
